text/0762-serverless-elasticache-l2-construct.md

```diff
@@ -0,0 +1,836 @@
+
+```ts
+/**
+ * Represents a Serverless ElastiCache cache
+ */
+export interface IServerlessCache extends IResource, ec2.IConnectable {
+  /**
+   * The cache engine used by this cache
+   */
+  readonly engine?: CacheEngine;
+  /**
+   * The name of the serverless cache
+   *
+   * @attribute
+   */
+  readonly serverlessCacheName: string;
+  /**
+   * The ARNs of backups restored in the cache
+   */
+  readonly backupArnsToRestore?: string[];
+  /**
+   * The KMS key used for encryption
+   */
+  readonly kmsKey?: kms.IKey;
+  /**
+   * The VPC this cache is deployed in
+   */
+  readonly vpc?: ec2.IVpc;
+  /**
+   * The subnets this cache is deployed in
+   */
+  readonly subnets?: ec2.ISubnet[];
+  /**
+   * The security groups associated with this cache
+   */
+  readonly securityGroups?: ec2.ISecurityGroup[];
+  /**
+   * The user group associated with this cache
+   */
+  readonly userGroup?: IUserGroup;
+  /**
+   * The ARN of the serverless cache
+   *
+   * @attribute
+   */
+  readonly serverlessCacheArn: string;
+
+  /**
+   * Grant connect permissions to the cache
+   */
+  grantConnect(grantee: iam.IGrantable): iam.Grant;
+  /**
+   * Grant the given identity custom permissions
+   */
+  grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant;
+
+  /**
+   * Return the given named metric for this cache
+   */
+  metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for cache hit count
+   */
+  metricCacheHitCount(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for cache miss count
+   */
+  metricCacheMissCount(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for cache hit rate
+   */
+  metricCacheHitRate(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for data stored in the cache
+   */
+  metricDataStored(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for ECPUs consumed
+   */
+  metricECPUsConsumed(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for network bytes in
+   */
+  metricNetworkBytesIn(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for network bytes out
+   */
+  metricNetworkBytesOut(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for active connections
+   */
+  metricActiveConnections(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for write request latency
+   */
+  metricWriteRequestLatency(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
+   * Metric for read request latency
+   */
+  metricReadRequestLatency(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+}
+```
+
+#### User
+
+The user authentication system has three types (IAM, Password, and No-Password),
+with each type implemented as a separate construct extending the UserBase class.
+
+Required fields examples:
+
+```ts
+new IamUser(this, 'IamUser', {
```

Comment:
it would be better to hide these classes from the user, and just expose some static factory methods.